File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.py
from random import choice


class RandomizedSet:
    def __init__(self):
        self.d = dict()
        self.l = list()


    def insert(self, val: int) -> bool:
        if val in self.d: return False
        self.d[val] = len(self.l)
        self.l.append(val)
        return True


    def remove(self, val: int) -> bool:
        if val not in self.d: return False
        index, last = self.d[val], self.l[-1]
        self.d[last] = index
        self.l[index] = last
        self.d.pop(val)
        self.l.pop()
        return True


    def getRandom(self) -> int:
        return choice(self.l)

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;


class RandomizedSet {
    Random random;
    HashMap<Integer, Integer> dict;
    ArrayList<Integer> list;

    public RandomizedSet() {
        dict = new HashMap<>();
        list = new ArrayList<>();
        random = new Random();
    }
    
    public boolean insert(int val) {
        if (dict.containsKey(val)) return false;
        dict.put(val, list.size());
        list.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if (!dict.containsKey(val)) return false;
        int index = dict.get(val);
        int last = list.get(list.size() - 1);
        dict.put(last, index);
        list.set(index, last);
        dict.remove(val);
        list.remove(list.size() - 1);
        return true;
    }
    
    public int getRandom() {
        return list.get(random.nextInt(list.size()));
    }
}
